packages/database/src/migrations/20231120003955-CreateDuplicateSurveys-modifies-data.js

```diff
@@ -0,0 +1,234 @@
+
+  const existingSurvey = await getExistingSurvey(db, oldSurveyCode);
+
+  // if that survey doesn't exist, don't break, just return early
+  if (!existingSurvey) return;
+
+  const {
+    id,
+    name,
+    can_repeat,
+    survey_group_id,
+    integration_metadata,
+    requires_approval,
+    data_group_id,
+  } = existingSurvey;
+
+  const existingSurveyScreens = await getSurveyScreens(db, id);
+
+  const existingSurveyScreenComponents = await getSurveyScreenComponents(db, existingSurveyScreens);
+
+  const newSurveyId = generateId();
+
+  // create the data group
+  const dataGroup = await getDataGroup(db, data_group_id);
+  const newSurveyDataGroupId = generateId();
+  await insertObject(db, 'data_group', {
+    ...dataGroup,
+    id: newSurveyDataGroupId,
+    code: newSurveyCode,
+  });
+
+  // create the survey
+  await insertObject(db, 'survey', {
+    id: newSurveyId,
+    code: newSurveyCode,
+    country_ids: `{${countryId}}`,
+    project_id: newProjectId,
+    permission_group_id: permissionGroupId,
+    name: getNewSurveyName(name),
+    can_repeat,
+    survey_group_id,
+    integration_metadata,
+    requires_approval,
+    data_group_id: newSurveyDataGroupId,
+  });
+
+  // map the old survey screen ids to new survey screen ids
+  const surveyScreenIdMap = existingSurveyScreens.reduce(
+    (acc, screen) => ({ ...acc, [screen.id]: generateId() }),
+    {},
+  );
+
+  // create the new survey screens
+  await Promise.all(
+    existingSurveyScreens.map(screen =>
+      insertObject(db, 'survey_screen', {
+        screen_number: screen.screen_number,
+        id: surveyScreenIdMap[screen.id],
+        survey_id: newSurveyId,
+      }),
+    ),
+  );
+
+  // create the new survey screen components
+  const newSurveyScreenComponents = existingSurveyScreenComponents.map(component => {
+    const {
+      question_id,
+      component_number,
+      is_follow_up,
+      visibility_criteria,
+      config,
+      validation_criteria,
+    } = component;
+    const newSurveyScreenId = surveyScreenIdMap[component.screen_id];
+    return {
+      question_id,
+      component_number,
+      is_follow_up,
+      id: generateId(),
+      screen_id: newSurveyScreenId,
+      visibility_criteria,
+      config,
+      validation_criteria,
+    };
+  });
+
+  await Promise.all(
+    newSurveyScreenComponents.map(component =>
+      insertObject(db, 'survey_screen_component', component),
+    ),
+  );
+};
+
+exports.up = async function (db) {
+  // all the duplicated surveys will be in the same country, Demo Land
+  const country = await db.runSql(`SELECT id FROM country WHERE name = 'Demo Land'`);
+
+  const permissionGroup = await db.runSql(`SELECT id FROM permission_group WHERE name = 'Public'`);
+  return Promise.all(
+    surveys.map(survey =>
+      createNewSurvey(db, survey, country.rows[0].id, permissionGroup.rows[0].id),
+    ),
+  );
+};
+
+const removeSurvey = async (db, survey) => {
```

Comment:
Maybe worth adding a `survey_response` deletion to this too? Just in case we want to undo this migration after a survey has been submitted?